Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,664 @@
+     // Used to re-trigger the verification of some diagnostics.
+     string Identifier,
+     string Filename,
+     // The range of this node.
+     Range Range
+  ) {
+    // Overriden by checking children if there are some
+    public VerificationStatus StatusVerification { get; set; } = VerificationStatus.Unknown;
+
+    // Overriden by checking children if there are some
+    public CurrentStatus StatusCurrent { get; set; } = CurrentStatus.Obsolete;
+
+    // Used to relocate a node diagnostic and to determine which function is currently verifying
+    public Position Position => Range.Start;
+
+    /// Time and Resource diagnostics
+    public bool Started { get; set; } = false;
+    public bool Finished { get; set; } = false;
+    public DateTime StartTime { get; protected set; }
+    public DateTime EndTime { get; protected set; }
+    public int TimeSpent => (int)(Finished ? ((TimeSpan)(EndTime - StartTime)).TotalMilliseconds : Started ? (DateTime.Now - StartTime).TotalMilliseconds : 0);
+    // Resources allocated at the end of the computation.
+    public int ResourceCount { get; set; } = 0;
+
+
+
+    // If this node is an error, all the trace positions
+    public ImmutableList<Range> RelatedRanges { get; set; } = ImmutableList<Range>.Empty;
+
+    // Sub-diagnostics if any
+    public List<NodeDiagnostic> Children { get; set; } = new();
+    private List<NodeDiagnostic> NewChildren { get; set; } = new();
+
+    public int GetNewChildrenCount() {
+      return NewChildren.Count;
+    }
+
+    public void AddNewChild(NodeDiagnostic newChild) {
+      NewChildren.Add(newChild);
+    }
+
+    public void SaveNewChildren() {
+      Children = NewChildren;
+      ResetNewChildren();
+    }
+    public void ResetNewChildren() {
+      NewChildren = new();
+    }
+
+    public NodeDiagnostic SetObsolete() {
+      if (StatusCurrent != CurrentStatus.Obsolete) {
+        StatusCurrent = CurrentStatus.Obsolete;
+        foreach (var child in Children) {
+          child.SetObsolete();
+        }
+      }
+
+      return this;
+    }
+
+    // Returns true if it started the method, false if it was already started
+    public virtual bool Start() {
+      if (StatusCurrent != CurrentStatus.Verifying || !Started) {
+        StartTime = DateTime.Now;
+        StatusCurrent = CurrentStatus.Verifying;
+        foreach (var child in Children) {
+          child.Start();
+        }
+        Started = true;
+        return true;
+      }
+
+      return false;
+    }
+
+    // Returns true if it did stop the current node, false if it was already stopped
+    public virtual bool Stop() {
+      if (StatusCurrent != CurrentStatus.Current || !Finished) {
+        EndTime = DateTime.Now;
+        StatusCurrent = CurrentStatus.Current;
+        foreach (var child in Children) {
+          child.Stop();
+        }
+        Finished = true;
+        return true;
+      }
+
+      return false;
+    }
+
+    public void PropagateChildrenErrorsUp() {
```

Comment:
Before sending diagnostics, I do this bottom-up propagation of errors to ensure that, if there is a descendant of that node with one error, this node is marked with error.